Shopify/cli
# Dynamic schema from contract

### WHY are these changes introduced?

Adds support for app config modules that are primarily validated by JSON Schema based contracts. Our first such module is used for custom data definitions.



### WHAT is this pull request doing?



Commit-by-commit is probably a good way to go through this, as there's some tidying happening too:
- `SpecsAppConfiguration` is no more, and we add `AppConfigurationUsedByCli`, which is a much narrower interface
- Extension specification schemas are no longer exposed, and methods provided instead to validate
- The loader function(s) are now more of a pipeline, which an app conifguration state method available, providing _just enough_ information to e.g. create a dev platform client, prior to loading the app proper. This is important: we need apps to only load having gone through a specification fetch.
- A JSON schema validator is added, with some support for matching Zod behaviour
- An error simplifier for union-style errors is added. It works well for structures that have e.g. a `type` key that acts as a discriminator
- Support for dynamic config entries via an env variable is dropped
- A new factory for modules that are contract based is added

### How to test your changes?



Add:
```
[metaobjects.author.fields]
name = "single_line_text_field"
```
to your config.

You should be able to deploy and `app info --json` this (the latter should have a data section), you should also be able to `link` and pull in the deployed config into the appropriate format.

Testing the contracts themselves is not currently possible as they're not available on the platform _just yet_.

### Post-release steps



### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
